### PR TITLE
Enable secure Jenkins build control for Windows

### PR DESCRIPTION
**Pull request by zanderz**
@jzila let me add a slackbot account on Jenkins, the credentials of which must be set in the environment with JENKINS_WINDOWS_USERNAME and JENKINS_WINDOWS_PASSWORD.

This makes building and cancelling work, but I'm not sure about the other commands. Would it be better to have a winbot? 

Also still figuring out how to have Jenkins notify Slack via groovy script.

@gabriel @chrisnojima

### CODE REVIEW

**Comment by jzila** [comment]
@zanderz: https://github.com/keybase/jenkins-helpers/blob/master/helpers.groovy#L103

File: keybot/keybot.go [hunk 1]
```diff
@@ -67,8 +74,15 @@ func (k *keybot) Run(bot slackbot.Bot, channel string, args []string) (string, e
 	env := launchd.NewEnv(home, path)
 	switch cmd {
 	case cancel.FullCommand():
+		if *cancelWin {
+			jenkins.StopBuild(*cancelLabel)
+			out := "Issued cancel"
+			if *cancelLabel != "" {
```
Comment (gabriel) [inline keybot/keybot.go]: Is cancelLabel required? If nothing happens when you don't specify, return an error?
```go
if *cancelLabel == "" {
  return nil, fmt.Errorf("Need to specify job name (label)")
}
```

File: keybot/keybot.go [hunk 2]
```diff
@@ -158,7 +172,12 @@ func (k *keybot) Run(bot slackbot.Bot, channel string, args []string) (string, e
 			},
 		}
 		return runScript(bot, channel, env, script)
+
+	// Windows
```
Comment (gabriel) [inline keybot/keybot.go]: Can remove this comment.

File: jenkins/jenkins.go [hunk 3]
```diff
@@ -138,11 +250,31 @@ func getLastBuildAndQueueNumbers() (currentBuild int, currentQ int, err error) {
 	return
 }
 
-// StopBuild will stop/cancel the current build.
+func controlJob(enable bool) {
```
Comment (gabriel) [inline jenkins/jenkins.go]: Probably want this to return error.

File: jenkins/jenkins.go [hunk 4]
```diff
@@ -98,7 +210,7 @@ func cancelQueueItem(queueEntry string) {
 	u.RawQuery = urlValues.Encode()
 	cancelURL := u.String()
```
Comment (gabriel) [inline jenkins/jenkins.go]: cancelQueueItem should return error?

File: jenkins/jenkins.go [hunk 5]
```diff
@@ -59,14 +168,17 @@ func StartBuild(clientRev string, kbfsRev string, jsonUpdateFilename string) (st
...
 	loc, _ := res.Location()
-	log.Print(robots)
+	log.Printf("StartBuild robots: %v", robots)
+
+	controlJob(true)
+
 	return fmt.Sprintf("Requested Jenkins build with queue ID %s", parseQueueNumber(loc.String())), nil
 }
 
 func stopBuildByID(buildID string) {
```
Comment (gabriel) [inline jenkins/jenkins.go]: should return error?

File: jenkins/jenkins.go [hunk 6]
```diff
@@ -138,11 +250,31 @@ func getLastBuildAndQueueNumbers() (currentBuild int, currentQ int, err error) {
...
+	if err != nil {
+		log.Print(err)
+	}
+}
+
+// StopBuild will disable auto builds and
+// stop/cancel the current build, if specified.
 // Location is the return string from a build request, e.g.:
 // http://192.168.1.10:8080/queue/item/34/
 func StopBuild(queueEntry string) {
```
Comment (gabriel) [inline jenkins/jenkins.go]: Should return error

**Pull request merged**